File: backend/src/main/java/com/team05/petmeeting/domain/user/controller/UserAuthController.java
package com.team05.petmeeting.domain.user.controller;

import com.team05.petmeeting.domain.user.dto.emailsignup.EmailSignupReq;
import com.team05.petmeeting.domain.user.dto.emailstart.EmailStartReq;
import com.team05.petmeeting.domain.user.dto.emailstart.EmailStartRes;
import com.team05.petmeeting.domain.user.dto.emailverify.EmailVerifyReq;
import com.team05.petmeeting.domain.user.dto.emailverify.EmailVerifyRes;
import com.team05.petmeeting.domain.user.dto.login.AccessTokenRes;
import com.team05.petmeeting.domain.user.dto.login.LoginAndRefreshResult;
import com.team05.petmeeting.domain.user.dto.login.local.EmailLoginReq;
import com.team05.petmeeting.domain.user.service.UserAuthService;
import com.team05.petmeeting.global.security.userdetails.CustomUserDetails;
import com.team05.petmeeting.global.security.util.RefreshTokenUtil;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/v1/auth")
@RequiredArgsConstructor
@Validated
public class UserAuthController {

    private final UserAuthService userAuthService;
    private final RefreshTokenUtil refreshTokenUtil;

    @PostMapping("/email/start")
    public ResponseEntity<EmailStartRes> startEmail(
            @RequestBody @Valid EmailStartReq request
    ) {
        EmailStartRes res = userAuthService.startEmailFlow(request.email());
        return ResponseEntity.ok(res);
    }

    @PostMapping("/email/send-otp")
    public ResponseEntity<Void> sendEmail(
            @RequestBody @Valid EmailStartReq request
    ) {
        userAuthService.sendSignupOtp(request.email());
        return ResponseEntity.noContent().build();
    }

    @PostMapping("/email/verify")
    public ResponseEntity<EmailVerifyRes> verifyEmail(
            @Valid @RequestBody EmailVerifyReq request
    ) {
        String verifyToken = userAuthService.verifyOtp(request.email(), request.code());
        return ResponseEntity.ok(new EmailVerifyRes(verifyToken));
    }

    @PostMapping("/email/signup")
    public ResponseEntity<AccessTokenRes> signupWithEmail(
            @Valid @RequestBody EmailSignupReq request,
            HttpServletResponse response
    ) {
        LoginAndRefreshResult result = userAuthService.signupAndLoginWithEmail(request);

        refreshTokenUtil.add(response, result.refreshToken());

        return ResponseEntity.status(HttpStatus.CREATED)
                .body(result.accessTokenRes());
    }

    @PostMapping("/email/login")
    public ResponseEntity<AccessTokenRes> loginWithEmail(
            @Valid @RequestBody EmailLoginReq request,
            HttpServletResponse response
    ) {
        LoginAndRefreshResult result = userAuthService.loginWithEmail(
                request.email(),
                request.password()
        );

        refreshTokenUtil.add(response, result.refreshToken());

        return ResponseEntity.ok(result.accessTokenRes());
    }

    // 로그아웃
    @PostMapping("/logout")
    public ResponseEntity<Void> logout(
            HttpServletRequest request,
            HttpServletResponse response
    ) {
        userAuthService.logout(request);
        refreshTokenUtil.delete(response);

        return ResponseEntity.noContent().build();
    }

    // 리프레시 토큰 재발급
    @PostMapping("/refresh")
    public ResponseEntity<AccessTokenRes> refresh(
            HttpServletRequest request,
            HttpServletResponse response
    ) {
        LoginAndRefreshResult result = userAuthService.refresh(request);

        // refresh token 재설정 (rotate)
        refreshTokenUtil.add(response, result.refreshToken());

        return ResponseEntity.ok(result.accessTokenRes());
    }

    // 탈퇴
    @DeleteMapping("/withdraw")
    public ResponseEntity<Void> withdraw(
            @AuthenticationPrincipal CustomUserDetails userDetails,
            HttpServletResponse response
    ) {
        userAuthService.withdraw(userDetails.getUserId());

        // refresh token 쿠키 제거
        refreshTokenUtil.delete(response);

        return ResponseEntity.noContent().build();
    }


}

File: backend/src/main/java/com/team05/petmeeting/domain/user/controller/UserAuthController.kt
package com.team05.petmeeting.domain.user.controller

import com.team05.petmeeting.domain.user.dto.auth.emailsignup.EmailSignupReq
import com.team05.petmeeting.domain.user.dto.auth.emailstart.EmailStartReq
import com.team05.petmeeting.domain.user.dto.auth.emailstart.EmailStartRes
import com.team05.petmeeting.domain.user.dto.auth.emailverify.EmailVerifyReq
import com.team05.petmeeting.domain.user.dto.auth.emailverify.EmailVerifyRes
import com.team05.petmeeting.domain.user.dto.auth.login.AccessTokenRes
import com.team05.petmeeting.domain.user.dto.auth.login.LoginAndRefreshRes
import com.team05.petmeeting.domain.user.dto.auth.login.local.EmailLoginReq
import com.team05.petmeeting.domain.user.service.UserAuthService
import com.team05.petmeeting.global.security.userdetails.CustomUserDetails
import com.team05.petmeeting.global.security.util.RefreshTokenUtil
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/v1/auth")
@Validated
class UserAuthController(
    private val userAuthService: UserAuthService,
    private val refreshTokenUtil: RefreshTokenUtil,
) {

    @PostMapping("/email/start")
    fun startEmail(
        @Valid @RequestBody request: EmailStartReq,
    ): ResponseEntity<EmailStartRes> =
        userAuthService.startEmailFlow(request.email)
            .let { ResponseEntity.ok(it) }

    @PostMapping("/email/send-otp")
    fun sendEmail(
        @Valid @RequestBody request: EmailStartReq,
    ): ResponseEntity<Void> =
        userAuthService.sendSignupOtp(request.email)
            .let { noContent() }

    @PostMapping("/email/verify")
    fun verifyEmail(
        @Valid @RequestBody request: EmailVerifyReq,
    ): ResponseEntity<EmailVerifyRes> =
        userAuthService.verifyOtp(request.email, request.code)
            .let(::EmailVerifyRes)
            .let { ResponseEntity.ok(it) }

    @PostMapping("/email/signup")
    fun signupWithEmail(
        @Valid @RequestBody request: EmailSignupReq,
        response: HttpServletResponse,
    ): ResponseEntity<AccessTokenRes> =
        userAuthService.signupAndLoginWithEmail(request)
            .withRefreshTokenCookie(response)
            .let { ResponseEntity.status(HttpStatus.CREATED).body(it) }

    @PostMapping("/email/login")
    fun loginWithEmail(
        @Valid @RequestBody request: EmailLoginReq,
        response: HttpServletResponse,
    ): ResponseEntity<AccessTokenRes> =
        userAuthService.loginWithEmail(request.email, request.password)
            .withRefreshTokenCookie(response)
            .let { ResponseEntity.ok(it) }

    @PostMapping("/logout")
    fun logout(
        request: HttpServletRequest,
        response: HttpServletResponse,
    ): ResponseEntity<Void> =
        userAuthService.logout(request)
            .also { refreshTokenUtil.delete(response) }
            .let { noContent() }

    @PostMapping("/refresh")
    fun refresh(
        request: HttpServletRequest,
        response: HttpServletResponse,
    ): ResponseEntity<AccessTokenRes> =
        userAuthService.refresh(request)
            .withRefreshTokenCookie(response)
            .let { ResponseEntity.ok(it) }

    @DeleteMapping("/withdraw")
    fun withdraw(
        @AuthenticationPrincipal userDetails: CustomUserDetails?,
        response: HttpServletResponse,
    ): ResponseEntity<Void> =
        userDetails.requireUserId()
            .also(userAuthService::withdraw)
            .also { refreshTokenUtil.delete(response) }
            .let { noContent() }

    private fun LoginAndRefreshRes.withRefreshTokenCookie(response: HttpServletResponse): AccessTokenRes =
        also { refreshTokenUtil.add(response, it.refreshToken) }
            .accessTokenRes

    private fun CustomUserDetails?.requireUserId(): Long =
        requireNotNull(this) { "인증된 사용자 정보가 없습니다." }
            .userId

    private fun noContent(): ResponseEntity<Void> =
        ResponseEntity.noContent().build()
}
